### PR TITLE
Fix several issues with the hashing method

### PR DESCRIPTION
**Pull request by dev-zero**
This is an an alternative to #2049 using Blake2b as the hash algorithm (in tree configuration).

Besides fixing the same issues, I discovered some other problems which slipped through the cracks:

* the dispatch on `int` type works differently between Python 2 and 3: while a `np.int64` was matched by it in Python 2, Python 3 was using the string type instead, which is why we are using `numbers.Integral` now
*  we can't distinguish in hash types between str, bytes and unicode. Otherwise a simple `"..."` will be a differerent hash type in Python 2 and 3
* furthermore I also made everything byte-ordner agnostic by converting numerical types to little-endian if necessary
* for some of the representations I used `struct.pack` with the little-endian standard for getting the byte representation instead of the string representation. Arbitrary length of integers are therefore currently not supported I guess.

I guess it may make sense to squash @greschd 's commits with my main commit. So, instead of reviewing the diff, it is probably better to review the final code.

### CODE REVIEW

File: aiida/common/hashing.py [hunk 1]
```diff
@@ -156,192 +150,193 @@ def make_hash(object_to_hash, **kwargs):
...
+        for name in sorted(subfolder.get_content_list()):
+            if name in ignored_folder_content:
+                continue
+
+            if subfolder.isdir(name):
+                yield _single_digest('folder', name.encode('utf-8'))
+                for digest in folder_digests(subfolder.get_subfolder(name)):
+                    yield digest
+            else:
+                with subfolder.open(name, mode='rb') as fhandle:
```
Comment (giovannipizzi) [inline aiida/common/hashing.py]: Missing hash of the filename

File: aiida/common/hashing.py [hunk 2]
```diff
@@ -156,192 +150,193 @@ def make_hash(object_to_hash, **kwargs):
...
+    If the type is unicode in Python 2 or a str in Python 3, convert it
+    to a str in Python 2 and bytes in Python 3 using the utf-8 encoding.
+    """
+    return [_single_digest('str', val.encode('utf-8'))]
+
+
+@_make_hash.register(abc.Sequence)
+def _(sequence_obj, **kwargs):
+    # unpack the list and use the elements
+    return [_single_digest('list')] + list(chain.from_iterable(_make_hash(i, **kwargs) for i in sequence_obj))
```
Comment (giovannipizzi) [inline aiida/common/hashing.py]: One needs a end of sequence (same for sets, dicts)

File: aiida/common/hashing.py [hunk 3]
```diff
@@ -156,192 +150,193 @@ def make_hash(object_to_hash, **kwargs):
...
+                continue
+
+            if subfolder.isdir(name):
+                yield _single_digest('folder', name.encode('utf-8'))
+                for digest in folder_digests(subfolder.get_subfolder(name)):
+                    yield digest
+            else:
+                with subfolder.open(name, mode='rb') as fhandle:
+                    yield _single_digest('file', fhandle.read())
+
```
Comment (giovannipizzi) [inline aiida/common/hashing.py]: yield a end-of-folder here

File: aiida/common/hashing.py [hunk 4]
```diff
@@ -156,192 +150,193 @@ def make_hash(object_to_hash, **kwargs):
...
-            ], **kwargs).encode('latin1')
-        )
-    else:
-        return make_hash_with_type(
-            'ao',
-            make_hash(object_to_hash.tobytes(), **kwargs).encode('latin1')
-        )
+    def folder_digests(subfolder):
+        """traverses the given folder and yields digests for the contained objects"""
+        for name in sorted(subfolder.get_content_list()):
```
Comment (giovannipizzi) [inline aiida/common/hashing.py]: Todo: Ensure this function consistently returns unicode

File: aiida/common/test_hashing.py [hunk 5]
```diff
@@ -32,72 +54,111 @@ class MakeHashTest(unittest.TestCase):
...
+        self.assertEqual(
+            make_hash({
+                'a': 'b',
+                'c': 'd'
+            }), '1890162a70dbbcbef2d0cc40b9bd90463459f2e7e72573628e7f1c003ca25c95')
+
+        # order changes in dictionaries should give the same hashes
+        self.assertEqual(
+            make_hash(collections.OrderedDict([('c', 'd'), ('a', 'b')])),
+            make_hash(collections.OrderedDict([('a', 'b'), ('c', 'd')])))
```
Comment (giovannipizzi) [inline aiida/common/test_hashing.py]: Suggestion: for make_hash, when an OrderedDict is passed, check for a kwarg (e.g. 'consider_as_unsorted' or a better name): if True, continue; if False, raise

**Comment by dev-zero** [comment]
@giovannipizzi do you want to talk another look at the changes https://github.com/aiidateam/aiida_core/pull/2110/commits/dec77bed8f0b41401f00fdb5d8a13758dac14710 ?

**Comment by dev-zero** [comment]
@giovannipizzi I rebased, can you do a squash&merge please?

**Pull request merged**